proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,329 @@
+
+- Up-to-date algorithms
+- Hierarchy of typeclasses
+- Monadic interfaces
+- Safety abstractions
+- Type-level composition
+- Advanced data structure (eg, merkle trees)
+- Use of Generic for crypto-serialization
+
+As a result, there is much to be gained from investing in foundational Haskell cryptography. Haskell is well-suited to providing many aspects of information security through use of monadic interfaces and the type system, if the lower-level challenges of strictness and performant primitives can be solved. The community needs a cryptography library that is easily maintained, with minimal dependencies / baggage, and with documentation, tutorials, and unit tests to ease development.
+
+A first step would be developing bindings to a suitably-licensed open-source C or C++ library containing modern and up-to-date cryptography algorithms, including post-quantum cryptography schemes.
+
+There are [several C/C++ library candidates](https://en.wikipedia.org/wiki/Comparison_of_cryptography_libraries):
+
+- Botan
+- libgcrypt
+- Crypto++
+- libsodium / NaCl
+- NSS
+
+Of these, Botan was selected for several reasons:
+
+- It has a BSD license
+- It has a C FFI
+- It offers a broad variety of functionality and algorithms, including post-quantum cryptography
+- It has been [audited](https://botan.randombit.net/releases/audit_1.11.18.pdf) in the past
+- There are dead bindings targeting Botan 2 / GHC 8 that can be examined / revived
+- Significant implementation work has already been performed in order to validate the feasibility of Botan
+
+The others were not selected for several reasons:
+
+- libgcrypt is C so it requires no FFI, but it has a GNU LGPL license, and provides less functionality than Botan
+- Crypto++ has a Boost license (similar to BSD), but is C++ and has no C FFI, and provides less functionality than Botan
+- libsodium / NaCL does not provide the necessary functionality and there are several Haskell libsodium bindings already
+- NSS has a copyleft license
+
+Bindings to Botan would solve the significant problem of dealing with common cryptography by providing much of the necessary 'cryptographic kitchen sink' via bindings to a suitably performant, suitably licensed, open-source library. The Haskell community would not be required to maintain the Botan cryptography library, only bindings to it, and this can be accomplished readily with some effort which has already begun. Furthermore, the Botan library supports several post-quantum cryptography algorithms, which is highly desirable.
+
+It is important that the Haskell community has a strong cryptography library / presence that is not bent towards a particular agenda. It must be **free as in speech (licensing)** and **free as in beer (no mining)**, in order to allow the growth of advanced cryptographic techniques the way that Haskell has furthered the growth of functional programming techniques.
```

Comment:
I'm not sure I understand this paragraph. What sorts of agendas would one worry about in a hash function? And are we really worried about rogue `libsodium` bindings mining cryptocurrencies here? Furthermore, it seems to me that the strongest argument for this proposal is that boring old everyday crypto is missing, not that we need to innovate with advanced crypto in our programs. Innovation is good, but keeping things relevant to standard practice is more important.

I'm not sure this paragraph contributes much to the argument, really.

Reply:
I can strike this, it is not technically relevant.